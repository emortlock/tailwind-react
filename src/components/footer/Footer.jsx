import React from 'react'
import PropTypes from 'prop-types'

import { Base } from '../primitives'
import { withTheme } from '../theme'
import { Container } from '../container'

const Footer = ({ theme, is, children, ...rest }) => (
  <Base
    is={is}
    role="contentinfo"
    p={{ t: theme.spacing.lg, b: theme.spacing.xl }}
    bg={theme.brandColors.secondary}
    text={theme.textColors.on.secondary}
    {...rest}
  >
    <Container padding>{children}</Container>
  </Base>
)

Footer.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
}

Footer.defaultProps = {
  is: 'footer',
  children: undefined,
}

export { Footer as component }
export default withTheme(Footer)
