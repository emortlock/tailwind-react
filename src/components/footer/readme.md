Simple wrapper around the `<footer />` element

### Example

```jsx
<Footer>
  <Text is="p">&copy; E Corp {new Date().getFullYear()}</Text>
  <Text size="sm">
    Cookies are used by our fictional third party partners to track your
    browsing habits, DNA and thoughts. View our{' '}
    <a href="#footer" className="text-white hover:text-black">
      privacy policy
    </a>{' '}
    to attempt to manage your preferences.
  </Text>
</Footer>
```
