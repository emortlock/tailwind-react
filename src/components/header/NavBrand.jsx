import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { Base } from '../primitives'

const NavBrand = ({
  theme,
  header: { style, screen },
  is,
  children,
  ...rest
}) => {
  const responsive = screen
    ? {
        [`m-${screen}`]: { r: theme.spacing.lg },
      }
    : {}

  const aria = !(typeof is === 'string' && is.startsWith('h'))
    ? {
        role: 'heading',
        'aria-level': 1,
      }
    : {}

  return (
    <Base
      is={is}
      inlineBlock
      noUnderline
      flex={[true, 'no-shrink']}
      items="center"
      h={12}
      text={style.text || theme.textColors.on.primary}
      {...responsive}
      {...aria}
      {...rest}
    >
      {children}
    </Base>
  )
}

NavBrand.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
}

NavBrand.defaultProps = {
  is: 'div',
  children: undefined,
  header: { style: {}, screen: 'lg' },
}

export { NavBrand as component }
export default withTheme(NavBrand)
