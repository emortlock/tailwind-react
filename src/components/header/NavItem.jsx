import React from 'react'
import PropTypes from 'prop-types'

import { withTheme } from '../theme'
import { BaseComponent } from '../tailwind'

const NavItem = ({
  theme,
  is,
  children,
  header: { style },
  active,
  ...rest
}) => {
  const textColor = style.text || theme.textColors.on.primary
  const bgColor = style.bg || theme.brandColors.primary

  return (
    <BaseComponent
      is={is}
      focusable
      text={!active ? style.text : style.bg}
      bg={active ? textColor : undefined}
      bg-hocus={textColor}
      text-hocus={bgColor}
      p={{ x: theme.spacing.md, y: theme.spacing.sm }}
      m={{ t: theme.spacing.sm }}
      m-lg={{ t: 0, r: theme.spacing.sm }}
      rounded={theme.radius}
      noUnderline
      block
      aria-current={active ? 'page' : undefined}
      {...rest}
    >
      {children}
    </BaseComponent>
  )
}

NavItem.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    style: PropTypes.object.isRequired,
  }),
  active: PropTypes.bool,
}

NavItem.defaultProps = {
  is: 'a',
  children: undefined,
  header: { style: {} },
  active: false,
}

export { NavItem as component }
export default withTheme(NavItem)
