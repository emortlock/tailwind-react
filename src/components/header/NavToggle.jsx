import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Button } from '../button'
import { withTheme } from '../theme'

const Bar = () => (
  <span
    className="border-b inline-block"
    style={{ borderColor: 'currentColor' }}
  />
)

const NavToggle = ({
  theme,
  children,
  className,
  onClick,
  header: { onToggle, style, id },
  ...rest
}) => {
  const handleClick = e => {
    onToggle()
    if (onClick) onClick(e)
  }

  return (
    <Button
      w={12}
      h={12}
      p={0}
      className={classnames('block lg:hidden', className)}
      onClick={handleClick}
      aria-label="Open menu"
      aria-haspopup="true"
      aria-controls={`${id}-nav`}
      text={style.text}
      bg-hocus={style.text}
      text-hocus={style.bg}
      {...rest}
    >
      {children || (
        <span className="flex flex-col items-stretch justify-around h-full p-3">
          <Bar />
          <Bar />
          <Bar />
        </span>
      )}
    </Button>
  )
}

NavToggle.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
  className: PropTypes.string,
  header: PropTypes.shape({
    onToggle: PropTypes.func.isRequired,
  }).isRequired,
  onClick: PropTypes.func,
}

NavToggle.defaultProps = {
  children: undefined,
  className: undefined,
  onClick: undefined,
}

export { NavToggle as component }
export default withTheme(NavToggle)
