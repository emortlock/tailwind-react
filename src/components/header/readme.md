Example:

```jsx
<Header>
  <NavBrand is="a" href="#header" font="semibold" text={['white', 'xl']}>
    <svg
      className="fill-current h-8 w-8 mr-2"
      width="54"
      height="54"
      viewBox="0 0 54 54"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M13.5 22.1c1.8-7.2 6.3-10.8 13.5-10.8 10.8 0 12.15 8.1 17.55 9.45 3.6.9 6.75-.45 9.45-4.05-1.8 7.2-6.3 10.8-13.5 10.8-10.8 0-12.15-8.1-17.55-9.45-3.6-.9-6.75.45-9.45 4.05zM0 38.3c1.8-7.2 6.3-10.8 13.5-10.8 10.8 0 12.15 8.1 17.55 9.45 3.6.9 6.75-.45 9.45-4.05-1.8 7.2-6.3 10.8-13.5 10.8-10.8 0-12.15-8.1-17.55-9.45-3.6-.9-6.75.45-9.45 4.05z" />
    </svg>
    <span className="inline-block">Tailwind React UI</span>
  </NavBrand>
  <NavToggle />
  <NavMenu>
    <NavItem is="a" href="#header" active>
      Docs
    </NavItem>
    <NavItem is="a" href="#header">
      Examples
    </NavItem>
    <NavItem is="a" href="#header">
      Blog
    </NavItem>
    <OutlineButton border="white" text="white" text-hocus="blue">
      Download
    </OutlineButton>
  </NavMenu>
</Header>
```

Custom branding example:

```jsx
<Header bg="purple" text="white">
  <NavBrand is="a" href="#header" font="semibold" text={['white', 'xl']}>
    <svg
      className="fill-current h-8 w-8 mr-2"
      width="54"
      height="54"
      viewBox="0 0 54 54"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M13.5 22.1c1.8-7.2 6.3-10.8 13.5-10.8 10.8 0 12.15 8.1 17.55 9.45 3.6.9 6.75-.45 9.45-4.05-1.8 7.2-6.3 10.8-13.5 10.8-10.8 0-12.15-8.1-17.55-9.45-3.6-.9-6.75.45-9.45 4.05zM0 38.3c1.8-7.2 6.3-10.8 13.5-10.8 10.8 0 12.15 8.1 17.55 9.45 3.6.9 6.75-.45 9.45-4.05-1.8 7.2-6.3 10.8-13.5 10.8-10.8 0-12.15-8.1-17.55-9.45-3.6-.9-6.75.45-9.45 4.05z" />
    </svg>
    <span className="inline-block">Tailwind React UI</span>
  </NavBrand>
  <NavToggle />
  <NavMenu>
    <NavItem is="a" href="#header" active>
      Docs
    </NavItem>
    <NavItem is="a" href="#header">
      Examples
    </NavItem>
    <NavItem is="a" href="#header">
      Blog
    </NavItem>
    <OutlineButton border="white" text="white" text-hocus="blue">
      Download
    </OutlineButton>
  </NavMenu>
</Header>
```
