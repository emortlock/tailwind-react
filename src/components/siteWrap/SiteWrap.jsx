import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { Base } from '../primitives'
import Footer, { component as FooterComponent } from '../footer/Footer'
import { withTheme } from '../theme'

const SiteWrap = ({ is, theme, children, ...rest }) => {
  let footer

  React.Children.forEach(children, child => {
    const { type } = child
    if (
      type &&
      (type === Footer || type.displayName === FooterComponent.displayName)
    ) {
      footer = child
    }
  })

  return (
    <Base
      is={is}
      flex={[true, 'col']}
      minH="screen"
      leading="normal"
      font={theme.text.family.body}
      text={[theme.text.size.body[1], theme.textColors.body]}
      {...rest}
    >
      <div className="flex-auto flex-no-shrink">
        {React.Children.map(children, child => {
          if (child === footer) return false
          return child
        })}
      </div>
      <div
        className={classnames(
          'flex-auto flex-no-shrink flex-no-grow',
          `mt-${theme.spacing.lg}`,
        )}
      >
        {footer}
      </div>
    </Base>
  )
}

SiteWrap.propTypes = {
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  theme: PropTypes.shape({}).isRequired,
  children: PropTypes.node,
}

SiteWrap.defaultProps = {
  is: 'div',
  children: undefined,
}

export { SiteWrap as component }
export default withTheme(SiteWrap)
