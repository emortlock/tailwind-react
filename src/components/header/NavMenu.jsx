import React from 'react'
import PropTypes from 'prop-types'
import classnames from 'classnames'

import { withTheme } from '../theme'
import { Box } from '../primitives'
import { Title } from '../typography'
import { withTransition } from '../utils'
import { List } from '../list'

import NavItem from './NavItem'

const NavItemWrapper = props => <li role="none" {...props} />

const NavMenu = ({
  theme,
  transition,
  is,
  children,
  header,
  list,
  ...rest
}) => {
  const transitionStyles = {
    entering: { maxHeight: '0', position: 'absolute' },
    entered: { maxHeight: '100vh' },
  }
  const headingId = `${header.id}-menu`

  const responsive = header.screen
    ? {
        [`w-${header.screen}`]: 'auto',
        [`flex-${header.screen}`]: true,
      }
    : {}

  return (
    <Box
      is={is}
      overflow="hidden"
      w="full"
      flex="grow"
      items="center"
      h={!header.collapsable ? 12 : undefined}
      style={
        header.collapsable
          ? {
              transition: 'max-height 500ms',
              maxHeight: '0',
              ...transitionStyles[transition],
            }
          : undefined
      }
      id={`${header.id}-nav`}
      aria-labelledby={headingId}
      aria-expanded={header.collapsable ? header.open : undefined}
      role="navigation"
      {...responsive}
      {...rest}
    >
      <Title level={2} id={headingId} visuallyHidden>
        Site menu
      </Title>
      <List
        reset
        className={classnames(
          'flex-grow',
          header.screen && `${header.screen}:flex ${header.screen}:mb-0`,
        )}
        role="menu"
        listItemIs={NavItemWrapper}
        {...list}
      >
        {React.Children.map(
          children,
          child =>
            child.type === NavItem &&
            React.cloneElement(child, { header, role: 'menuitem' }),
        )}
      </List>
      {React.Children.map(children, child => child.type !== NavItem && child)}
    </Box>
  )
}

NavMenu.propTypes = {
  theme: PropTypes.shape({}).isRequired,
  transition: PropTypes.string,
  is: PropTypes.oneOfType([PropTypes.string, PropTypes.func, PropTypes.object]),
  children: PropTypes.node,
  header: PropTypes.shape({
    open: PropTypes.bool,
    collapsable: PropTypes.bool,
    screen: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  }),
  list: PropTypes.shape({}),
}

NavMenu.defaultProps = {
  is: 'nav',
  children: undefined,
  header: {
    open: false,
    collapsable: false,
    screen: 'lg',
  },
  transition: 'entering',
  list: {},
}

export { NavMenu as component }
export default withTheme(withTransition(NavMenu, { inState: 'header.open' }))
