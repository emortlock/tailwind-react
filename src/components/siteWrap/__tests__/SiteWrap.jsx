import React from 'react'
import { shallow } from 'enzyme'

import SiteWrap from '../SiteWrap'
import { Header } from '../../header'
import { Footer } from '../../footer'

const setup = (testProps = {}) => {
  const props = Object.assign(testProps)

  const wrapper = shallow(
    <SiteWrap {...props}>
      <Header>Header</Header>
      <p>Body</p>
      <Footer>Footer</Footer>
    </SiteWrap>,
  )

  return {
    props,
    wrapper,
  }
}

describe('SiteWrap', () => {
  it('renders matching snapshot', () => {
    const { wrapper } = setup()

    expect(wrapper).toMatchSnapshot()
  })
})
