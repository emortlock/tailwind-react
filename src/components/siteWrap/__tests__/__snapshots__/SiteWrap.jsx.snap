// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SiteWrap renders matching snapshot 1`] = `
<Base
  flex={
    Array [
      true,
      "col",
    ]
  }
  font="sans"
  is="div"
  leading="normal"
  minH="screen"
  text={
    Array [
      "base",
      "grey-darkest",
    ]
  }
>
  <div
    className="flex-auto flex-no-shrink"
  >
    <WithTheme(Header)
      key=".0"
    >
      Header
    </WithTheme(Header)>
    <p
      key=".1"
    >
      Body
    </p>
  </div>
  <div
    className="flex-auto flex-no-shrink flex-no-grow mt-8"
  >
    <WithTheme(Footer)>
      Footer
    </WithTheme(Footer)>
  </div>
</Base>
`;
